stablehlo/transforms/StablehloWrapInComposite.cpp

```diff
@@ -0,0 +1,216 @@
+
+  // Create the return operation, returning the results of the cloned op.
+  llvm::SmallVector<mlir::Value> results;
+  results.append(cloned_op->getResults().begin(),
+                 cloned_op->getResults().end());
+  builder.create<mlir::func::ReturnOp>(implFunc.getBody().getLoc(), results);
+
+  // Add the newly created function to the module's symbol table and make it
+  // private.
+  mlir::SymbolTable symbol_table(module);
+  implFunc.setPrivate();
+  symbol_table.insert(implFunc);
+
+  return implFunc;
+}
+
+// A ConversionPattern that matches any operation and rewrites it as a
+// stablehlo::CompositeOp. The original operation's functionality is
+// encapsulated within a newly created private function.
+class ConvertGenericOp : public ConversionPattern {
+ public:
+  explicit ConvertGenericOp(MLIRContext* ctx)
+      : ConversionPattern(MatchAnyOpTypeTag(), /*benefit=*/1, ctx) {}
+
+  LogicalResult matchAndRewrite(
+      Operation* op, ArrayRef<Value> operands,
+      ConversionPatternRewriter& rewriter) const override {
+    // get the enclosing module
+    auto module = op->getParentOfType<ModuleOp>();
+    MLIRContext* context = op->getContext();
+    if (module == nullptr) {
+      return rewriter.notifyMatchFailure(op, "Failed to find enclosing module");
+    }
+    auto implFunc = buildStableHLOCompositeImplFunc(module, op);
+    auto name = op->getName().getStringRef();
+
+    llvm::SmallVector<mlir::Value> compositeOperands(op->operand_begin(),
+                                                     op->operand_end());
+    auto compositeOp = rewriter.create<stablehlo::CompositeOp>(
+        op->getLoc(), op->getResultTypes(), compositeOperands, name,
+        DictionaryAttr::get(context, op->getAttrs()), implFunc.getSymName());
+    rewriter.replaceOp(op, compositeOp.getResults());
+    return success();
+  }
+};
+
+class StablehloWrapInCompositePass
+    : public impl::StablehloWrapInCompositePassBase<
+          StablehloWrapInCompositePass> {
+ public:
+  StablehloWrapInCompositePass()
+      : StablehloWrapInCompositePassBase<StablehloWrapInCompositePass>() {}
+  StablehloWrapInCompositePass(const StablehloWrapInCompositePassOptions& opts)
+      : StablehloWrapInCompositePassBase<StablehloWrapInCompositePass>(opts) {}
+  explicit StablehloWrapInCompositePass(
+      std::function<bool(Operation*)> opPredicate) {
+    this->opPredicate = opPredicate;
+  }
+  LogicalResult initialize(MLIRContext* context) override {
+    RewritePatternSet patterns_(context);
+    patterns_.add<ConvertGenericOp>(context);
+    patterns = std::move(patterns_);
+
+    if (!opPredicate) {
+      if (!opNamesOption.empty()) {
+        DenseSet<StringRef> opNames(opNamesOption.begin(), opNamesOption.end());
+        opPredicate = [opNames](Operation* op) {
+          return opNames.contains(op->getName().getStringRef());
+        };
+      } else {
+        opPredicate = [](Operation* op) { return false; };
```

Comment:
Can we move this to the default compiler ctor?

```
  StablehloWrapInCompositePass(const StablehloWrapInCompositePassOptions& opts)
      : StablehloWrapInCompositePassBase<StablehloWrapInCompositePass>(opts), opPredicate([](){...}) {}
```

Reply:
Yes, I tried to move it here, but the issue is that the context is notinitialized yet. I need the context to initialize the `OperationName` objects from the `opNamesOption`.

Currently I am wrapping the initialization logic in `void initializePredicateMap(MLIRContext* context, ArrayRef<std::string> opNames)` which is called from the pass initialize function.